pkg/controllers/clusterresourceplacement/placement_controller.go

```diff
@@ -333,3 +340,153 @@ func (r *Reconciler) updatePlacementAppliedCondition(placement *fleetv1alpha1.Cl
+		// It could happen when the controller just sets the latest label to false for the old snapshot, and fails to
+		// create a new policy snapshot.
+		// And then the customers revert back their policy to the old one again.
+		// In this case, the "latest" snapshot without isLatest label has the same policy hash as the current policy.
+
+		latestPolicySnapshot.Labels[fleetv1.IsLatestSnapshotLabel] = strconv.FormatBool(true)
+		if err := r.Client.Update(ctx, latestPolicySnapshot); err != nil {
+			klog.ErrorS(err, "Failed to set the isLatestSnapshot label to true", "clusterPolicySnapshot", klog.KObj(latestPolicySnapshot))
+			return ctrl.Result{}, err
+		}
+	}
+	// create clusterResourceSnapshot
+	// update the status based on the latestPolicySnapshot status
+	// update the status based on the work
+	return ctrl.Result{}, nil
+}
+
+// lookupLatestClusterPolicySnapshot finds the latest snapshots and its policy index.
+// There will be only one active policy snapshot if exists.
+// It first checks whether there is an active policy snapshot.
+// If not, it finds the one whose policyIndex label is the largest.
+// The policy index will always start from 0.
+// Return error when 1) cannot list the snapshots 2) there are more than one active policy snapshots 3) snapshot has the
+// invalid label value.
+// 2 & 3 should never happen.
+func (r *Reconciler) lookupLatestClusterPolicySnapshot(ctx context.Context, crp *fleetv1.ClusterResourcePlacement) (*fleetv1.ClusterPolicySnapshot, int, error) {
+	snapshotList := &fleetv1.ClusterPolicySnapShotList{}
+	latestSnapshotLabelMatcher := client.MatchingLabels{
+		fleetv1.CRPTrackingLabel:      crp.Name,
+		fleetv1.IsLatestSnapshotLabel: strconv.FormatBool(true),
+	}
+	crpKObj := klog.KObj(crp)
+	if err := r.Client.List(ctx, snapshotList, latestSnapshotLabelMatcher); err != nil {
+		klog.ErrorS(err, "Failed to list active clusterPolicySnapshots", "clusterResourcePlacement", crpKObj)
+		return nil, -1, err
+	}
+	if len(snapshotList.Items) == 1 {
+		policyIndex, err := parsePolicyIndexFromLabel(&snapshotList.Items[0])
+		if err != nil {
+			return nil, -1, err
+		}
+		return &snapshotList.Items[0], policyIndex, nil
+	} else if len(snapshotList.Items) > 1 {
+		// It means there are multiple active snapshots and should never happen.
+		err := fmt.Errorf("there are %d active clusterPolicySnapshots owned by clusterResourcePlacement %v", len(snapshotList.Items), crp.Name)
+		klog.ErrorS(err, "It should never happen", "clusterResourcePlacement", crpKObj)
+		// TODO(zhiying) emit error metrics or well defined logs
+		return nil, -1, controller.NewUnexpectedBehaviorError(err)
+	}
+	// When there are no active snapshots, find the one who has the largest policy index.
+	if err := r.Client.List(ctx, snapshotList, client.MatchingLabels{fleetv1.CRPTrackingLabel: crp.Name}); err != nil {
+		klog.ErrorS(err, "Failed to list all clusterPolicySnapshots", "clusterResourcePlacement", crpKObj)
+		return nil, -1, err
```

Comment:
as we discussed, this is a dependency (api server) failure. Since we created Unexpected error type, shall we create this too?

Reply:
sure, will address